File: src/main/java/org/vw/domain/Grid.java
package org.vw.domain;

import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
public class Grid {
    private final int maxX;
    private final int maxY;
    private List<CleaningRobot> robots;

    public Grid(int maxX, int maxY) {
        this.maxX = maxX;
        this.maxY = maxY;
        robots = new ArrayList<>();
    }

    public void addRobot(CleaningRobot robot) {
        robots.add(robot);
    }

    public boolean isCellAvailable(int x, int y) {
        return this.isInsideGrid(x, y) && !this.isCellOccupied(x, y);
    }

    private boolean isInsideGrid(int newX, int newY) {
        return newX >= 0 && newX < maxX && newY >= 0 && newY < maxY;
    }

    private boolean isCellOccupied(int x, int y) {
        for (CleaningRobot robot : robots) {
            if (robot.getPosition().getX() == x && robot.getPosition().getY() == y) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/java/org/vw/domain/Grid.kt
package org.vw.domain

class Grid(var maxX: Int, var maxY: Int) {
    var robots: MutableList<CleaningRobot> = ArrayList()

    fun addRobot(robot: CleaningRobot) {
        robots.add(robot)
    }

    fun isCellAvailable(x: Int, y: Int): Boolean {
        return this.isInsideGrid(x, y) && !this.isCellOccupied(x, y)
    }

    private fun isInsideGrid(newX: Int, newY: Int): Boolean {
        return (newX in 0..<maxX) && (newY in 0..<maxY)
    }

    private fun isCellOccupied(x: Int, y: Int): Boolean {
        for (robot in robots) {
            if (robot.position.x == x && robot.position.y == y) {
                return true
            }
        }
        return false
    }
}
